homeassistant/components/onkyo/config_flow.py

```diff
@@ -0,0 +1,246 @@
+
+        return self.async_show_form(
+            step_id="pick_device",
+            data_schema=vol.Schema({vol.Required(CONF_DEVICE): vol.In(devices_names)}),
+        )
+
+    async def async_step_import(self, user_input: dict[str, Any]) -> ConfigFlowResult:
+        """Import the yaml config."""
+
+        host = user_input.get(CONF_HOST)
+        if host is None:
+            return self.async_abort(reason="no_host_defined")
+
+        info = None
+        try:
+            receiver = eiscp.eISCP(host, CONF_PORT_DEFAULT)
+            info = receiver.info
+        finally:
+            if receiver:
+                receiver.disconnect()
+
+        if info is None:
+            # Info is None when connection fails
+            return self.async_abort(reason="cannot_connect")
+
+        unique_id = info[EISCP_IDENTIFIER]
+        await self.async_set_unique_id(unique_id, raise_on_progress=False)
+        self._abort_if_unique_id_configured()
+
+        user_input[CONF_MODEL] = info[EISCP_MODEL_NAME]
+        user_input[CONF_MAC] = unique_id
+
+        options = {}
+        options[CONF_MAXIMUM_VOLUME] = user_input[CONF_LEGACY_MAXIMUM_VOLUME]
+        options[CONF_RECEIVER_MAXIMUM_VOLUME] = user_input[
+            CONF_LEGACY_RECEIVER_MAXIMUM_VOLUME
+        ]
+        options[CONF_SOURCES] = user_input[CONF_SOURCES]
+
+        title = self._get_receiver_name(info[EISCP_MODEL_NAME], unique_id)
+
+        return self.async_create_entry(title=title, data=user_input, options=options)
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: ConfigEntry,
+    ) -> OptionsFlow:
+        """Return the options flow."""
+        return OnkyoOptionsFlowHandler(config_entry)
+
+
+class OnkyoOptionsFlowHandler(OptionsFlowWithConfigEntry):
+    """Handle an options flow for Onkyo."""
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Manage the options."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                SCHEMA_SOURCES = vol.Schema({str: str})
+                SCHEMA_SOURCES(user_input.get(CONF_SOURCES))
+            except vol.error.Invalid:
+                errors["base"] = "invalid_sources"
+
+            if not errors:
+                return self.async_create_entry(data=user_input)
+
+        options_schema = vol.Schema(
+            {
+                vol.Required(
+                    CONF_MAXIMUM_VOLUME, default=CONF_MAXIMUM_VOLUME_DEFAULT
+                ): vol.All(cv.positive_int, vol.Range(min=0, max=100)),
+                vol.Required(
+                    CONF_RECEIVER_MAXIMUM_VOLUME,
+                    default=CONF_RECEIVER_MAXIMUM_VOLUME_DEFAULT,
+                ): vol.All(cv.positive_int, vol.Range(min=0, max=200)),
+                vol.Required(
+                    CONF_SOURCES, default=CONF_SOURCES_DEFAULT
+                ): ObjectSelector(),
```

Comment:
There is no way to fetch all supported sources of a particular receiver and the list of is possibilities very long. This allows you to map from Onkyo setting (e.g. `tv/cd`, `video2`) to something that make sense in HA (e.g. `TV`, `chromecast`). This was ported from the original integration.

Reply:
Yea I did see it was in the old YAML, but I am curious how this looks like and how we can do this

Reply:
Hmm, so this leaves me with a double feeling, because with Home Assistant we want to represent the device itself as correct as possible without any calculated values in between. Even though calculated values would increase the usability and the user experience of that integration. Because when we look at the problem of "what device is `video2` actually?", that problem is a shared problem that isn't specific to Onkyo.

I am wondering, can we get the complete list of sources from the device, or should that always be set by the user to make sure its only limited to the sources we actually have?

I will discuss this with some others to have some more opinions on the table.

Reply:
> [...] only limited to the sources we actually have

This. To help the discussion let me give you the facts:

- The protocol allows for over 40 options to be sent as a `source`;
- The average receiver supports 10 or less;
- There is no command in the protocol to query a list of supported sources. 
   This holds true for `sound_mode` too, as well as any other settings that I wish to implement at a later date. 
- There is no friendly way to detect what settings are supported since it either:
  - is dependent on the current mode of the receiver;
  - will timeout (10s) per query if it is not supported
  - will change the receiver's mode if it is supported, making the device appear to go haywire.

The current Onkyo documentation even has you use the Python Onkyo tooling to obtain the current values. My goal is to add a service (different PR) that at least allows you to do it from the home assistant side.

Reply:
But in those cases it could be a combobox to be able to have the user select the sources
<img width="574" alt="image" src="https://github.com/home-assistant/core/assets/7083755/adae7008-70fb-4958-b282-521d667cefeb">

Reply:
I don't think it's very good UX to present over 40 options and I can't say for certain that any of the available source material is or will be exhaustive.

Next to that l, how would that interaction provide a way to enter a meaningful alias to the selected source? For instance in my case, video2 is setup to my chromecast, aux is playstation and cbl/sat is for Nintendo. I quite like the latter examples better 😊

What is it that you have against the dictionary approach?

If we want to have a more HA generic solution, an extendable key value list would make more sense in a lot of those cases imho.